In [89]:
from sklearn.preprocessing import StandardScaler
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.utils import resample
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score
import os






**Άσκηση 1 — Φόρτωση Δεδομένων**

Φορτώστε το αρχείο Excel. Τα δεδομένα περιλαμβάνουν:

• Χαρακτηριστικά: Στήλες A έως K (Δείκτες απόδοσης & δραστηριότητας).

• Στόχος (Target): Στήλη Status (1=Healthy, 2=Bankrupt).

• Αγνοήστε τη στήλη "Year"

In [79]:
df = pd.read_excel('Dataset2Use_Assignment.xlsx') 
df.head() #5 πρώτες εγραφές

365* ( Β.Υ / Κοστ.Πωλ )  Λειτ.Αποτ/Συν.Ενεργ. (ROA)  \
0              1014.128862                    0.011751   
1               561.974631                    0.099662   
2               826.919025                    0.009594   
3               493.176956                    0.196284   
4               100.421465                    0.179038   

   ΧΡΗΜ.ΔΑΠΑΝΕΣ / ΠΩΛΗΣΕΙΣ   ΠΡΑΓΜΑΤΙΚΗ ΡΕΥΣΤΟΤΗΤΑ :  (ΚΕ-ΑΠΟΘΕΜΑΤΑ) / Β.Υ  \
0                 0.038505                                        0.965479   
1                 0.015775                                        1.415467   
2                 0.093562                                        1.575034   
3                 0.000581                                        1.346780   
4                 0.000173                                        3.299687   

   (ΑΠΑΙΤ.*365) / ΠΩΛ.  Συν.Υποχρ/Συν.Ενεργ  Διάρκεια Παραμονής Αποθεμάτων  \
0           567.866550             0.532258                     545.195964   
1           335.917481             0.660556                       7.950910   
2           350.618086             0.550912                      28.602221   
3           414.508709             0.683717                       1.216888   
4           118.548381             0.254496                      52.346321   

   Λογαριθμος Προσωπικού  ΕΝΔΕΙΞΗ ΕΞΑΓΩΓΩΝ  ΕΝΔΕΙΞΗ ΕΙΣΑΓΩΓΩΝ  \
0               0.301030                 0                  1   
1               1.176091                 1                  1   
2               1.204120                 0                  1   
3               1.431364                 0                  0   
4               1.518514                 0                  1   

   ΕΝΔΕΙΞΗ ΑΝΤΙΠΡΟΣΩΠΕΙΩΝ  ΕΝΔΕΙΞΗ ΑΣΥΝΕΠΕΙΑΣ (=2) (ν+1)  ΕΤΟΣ  
0                       0                              1  2006  
1                       1                              1  2006  
2                       1                              1  2006  
3                       0                              1  2006  
4                       1                              1  2006

In [80]:
df.info() #10716 συνολικές εγγραφές


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10716 entries, 0 to 10715
Data columns (total 13 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   365* ( Β.Υ / Κοστ.Πωλ )                         10716 non-null  float64
 1   Λειτ.Αποτ/Συν.Ενεργ. (ROA)                      10716 non-null  float64
 2   ΧΡΗΜ.ΔΑΠΑΝΕΣ / ΠΩΛΗΣΕΙΣ                         10716 non-null  float64
 3    ΠΡΑΓΜΑΤΙΚΗ ΡΕΥΣΤΟΤΗΤΑ :  (ΚΕ-ΑΠΟΘΕΜΑΤΑ) / Β.Υ  10716 non-null  float64
 4   (ΑΠΑΙΤ.*365) / ΠΩΛ.                             10716 non-null  float64
 5   Συν.Υποχρ/Συν.Ενεργ                             10716 non-null  float64
 6   Διάρκεια Παραμονής Αποθεμάτων                   10716 non-null  float64
 7   Λογαριθμος Προσωπικού                           10716 non-null  float64
 8   ΕΝΔΕΙΞΗ ΕΞΑΓΩΓΩΝ                                10716 non-null  int64  
 9   ΕΝΔΕΙΞΗ ΕΙΣΑΓΩΓΩΝ                      

In [81]:
df = df.drop(columns=['ΕΤΟΣ']) #Διώχνουμε το  Έτος
df.head()

365* ( Β.Υ / Κοστ.Πωλ )  Λειτ.Αποτ/Συν.Ενεργ. (ROA)  \
0              1014.128862                    0.011751   
1               561.974631                    0.099662   
2               826.919025                    0.009594   
3               493.176956                    0.196284   
4               100.421465                    0.179038   

   ΧΡΗΜ.ΔΑΠΑΝΕΣ / ΠΩΛΗΣΕΙΣ   ΠΡΑΓΜΑΤΙΚΗ ΡΕΥΣΤΟΤΗΤΑ :  (ΚΕ-ΑΠΟΘΕΜΑΤΑ) / Β.Υ  \
0                 0.038505                                        0.965479   
1                 0.015775                                        1.415467   
2                 0.093562                                        1.575034   
3                 0.000581                                        1.346780   
4                 0.000173                                        3.299687   

   (ΑΠΑΙΤ.*365) / ΠΩΛ.  Συν.Υποχρ/Συν.Ενεργ  Διάρκεια Παραμονής Αποθεμάτων  \
0           567.866550             0.532258                     545.195964   
1           335.917481             0.660556                       7.950910   
2           350.618086             0.550912                      28.602221   
3           414.508709             0.683717                       1.216888   
4           118.548381             0.254496                      52.346321   

   Λογαριθμος Προσωπικού  ΕΝΔΕΙΞΗ ΕΞΑΓΩΓΩΝ  ΕΝΔΕΙΞΗ ΕΙΣΑΓΩΓΩΝ  \
0               0.301030                 0                  1   
1               1.176091                 1                  1   
2               1.204120                 0                  1   
3               1.431364                 0                  0   
4               1.518514                 0                  1   

   ΕΝΔΕΙΞΗ ΑΝΤΙΠΡΟΣΩΠΕΙΩΝ  ΕΝΔΕΙΞΗ ΑΣΥΝΕΠΕΙΑΣ (=2) (ν+1)  
0                       0                              1  
1                       1                              1  
2                       1                              1  
3                       0                              1  
4                       1                              1

In [82]:
counts = df['ΕΝΔΕΙΞΗ ΑΣΥΝΕΠΕΙΑΣ (=2) (ν+1)'].value_counts()
print("Counts:\n", counts)
ratio = counts[1] / counts[2]
print(f"\nRatio (Healthy/Bankrupt): {ratio:.2f}")

Counts:
 ΕΝΔΕΙΞΗ ΑΣΥΝΕΠΕΙΑΣ (=2) (ν+1)
1    10468
2      248
Name: count, dtype: int64

Ratio (Healthy/Bankrupt): 42.21


### Άσκηση 2 — Προεπεξεργασία
1. Κανονικοποίηση: Χρησιμοποιήστε StandardScaler στα χαρακτηριστικά.

In [83]:
scaler = StandardScaler()
X = df.drop(columns=['ΕΝΔΕΙΞΗ ΑΣΥΝΕΠΕΙΑΣ (=2) (ν+1)'])
X = pd.DataFrame(scaler.fit_transform(X), columns=X.columns)
X.head()


365* ( Β.Υ / Κοστ.Πωλ )  Λειτ.Αποτ/Συν.Ενεργ. (ROA)  \
0                 1.468457                   -0.203379   
1                 0.298940                    0.491425   
2                 0.984231                   -0.220431   
3                 0.120992                    1.255074   
4                -0.894888                    1.118765   

   ΧΡΗΜ.ΔΑΠΑΝΕΣ / ΠΩΛΗΣΕΙΣ   ΠΡΑΓΜΑΤΙΚΗ ΡΕΥΣΤΟΤΗΤΑ :  (ΚΕ-ΑΠΟΘΕΜΑΤΑ) / Β.Υ  \
0                 0.251958                                       -0.202928   
1                -0.346115                                        0.313641   
2                 1.700621                                        0.496818   
3                -0.745918                                        0.234791   
4                -0.756643                                        2.476655   

   (ΑΠΑΙΤ.*365) / ΠΩΛ.  Συν.Υποχρ/Συν.Ενεργ  Διάρκεια Παραμονής Αποθεμάτων  \
0             1.639054            -0.693019                       2.590355   
1             0.566376            -0.273256                      -0.598933   
2             0.634360            -0.631985                      -0.476339   
3             0.929831            -0.197480                      -0.638909   
4            -0.438875            -1.601793                      -0.335386   

   Λογαριθμος Προσωπικού  ΕΝΔΕΙΞΗ ΕΞΑΓΩΓΩΝ  ΕΝΔΕΙΞΗ ΕΙΣΑΓΩΓΩΝ  \
0              -1.292255         -0.789395           0.606966   
1               0.278482          1.266793           0.606966   
2               0.328793         -0.789395           0.606966   
3               0.736696         -0.789395          -1.647539   
4               0.893131         -0.789395           0.606966   

   ΕΝΔΕΙΞΗ ΑΝΤΙΠΡΟΣΩΠΕΙΩΝ  
0               -0.742307  
1                1.347151  
2                1.347151  
3               -0.742307  
4                1.347151

2. Κωδικοποίηση Στόχου: Μετατρέψτε το Status σε δυαδική μορφή (0=Healthy,
1=Bankrupt).

In [84]:
df = df.rename(columns={'ΕΝΔΕΙΞΗ ΑΣΥΝΕΠΕΙΑΣ (=2) (ν+1)': 'status'})
df['status'] = df['status'].map({1: 0, 2: 1})

3. Διαχωρισμός: Δημιουργήστε Train/Test sets (π.χ. test_size=0.25, stratified).

In [85]:
y = df['status']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, stratify=y, random_state=42)

print(f"Train shape: {X_train.shape}, Test shape: {X_test.shape}")

Train shape: (8037, 11), Test shape: (2679, 11)


**Άσκηση 3 — Έλεγχος Ισορροπίας (Class Balancing)**

Ελέγξτε την αναλογία Healthy:Bankrupt στο Training Set. Αν οι υγιείς επιχειρήσεις είναι
περισσότερες από 3 φορές τις χρεωκοπημένες (αναλογία > 3:1), κάντε downsampling (τυχαία
αφαίρεση) στην κλάση Healthy ώστε η αναλογία να γίνει ακριβώς 3:1.

In [86]:

#Γνωρίζουμε η ότι η αναλογία Healthy:Bankrupt είναι περίπου 42.21:1, άρα πρέπει να κάνουμε downsampling στην κλάση Healthy.

X_healthy = X_train[y_train == 0]
y_healthy = y_train[y_train == 0]
X_bankrupt = X_train[y_train == 1]
y_bankrupt = y_train[y_train == 1]
    
X_healthy_downsampled, y_healthy_downsampled = resample(
        X_healthy, y_healthy,
        replace=False,    
        n_samples=3 *(y_train == 1).sum(), 
        random_state=42
    )
    
X_train = pd.concat([X_healthy_downsampled, X_bankrupt])
y_train = pd.concat([y_healthy_downsampled, y_bankrupt])
    

print(f"Νέα Αναλογία Healthy: {(y_train == 0).sum()}, Bankrupt: {(y_train == 1).sum()}")
print(f"Ratio μετά το downsampling: {(y_train == 0).sum() / (y_train == 1).sum():.2f}")

Νέα Αναλογία Healthy: 558, Bankrupt: 186
Ratio μετά το downsampling: 3.00


**Άσκηση 4 — Εκπαίδευση Μοντέλων**

Εκπαιδεύστε και αξιολογήστε τους παρακάτω ταξινομητές:
• Linear Discriminant Analysis (LDA)
• Logistic Regression (max_iter=1000)
• K-Nearest Neighbors (KNN) (δοκιμάστε k=5, 7, 9)
• Naïve Bayes (GaussianNB)
• Support Vector Machines (SVM) (RBF kernel)
• Neural Networks (MLP) (64, 32 nodes, max_iter=400)

In [87]:
models = {
    "LDA": LinearDiscriminantAnalysis(),
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "KNN (k=5)": KNeighborsClassifier(n_neighbors=5),
    "KNN (k=7)": KNeighborsClassifier(n_neighbors=7),
    "KNN (k=9)": KNeighborsClassifier(n_neighbors=9),
    "Naïve Bayes": GaussianNB(),
    "SVM (RBF)": SVC(kernel='rbf'),
    "MLP (64, 32)": MLPClassifier(hidden_layer_sizes=(64, 32), max_iter=400, random_state=42)
}

# Train and evaluate
print(f"{'Model':<30} {'Accuracy':<10} {'Recall (Bankrupt)':<20} {'Specificity (Healthy)':<20} {'Constraints Met'}")
print("="*100)

for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    
    report = classification_report(y_test, y_pred, output_dict=True)
    recall_bankrupt = report['1']['recall']      # Sensitivity
    specificity = report['0']['recall']          # Specificity (Recall of Healthy class)
    acc = accuracy_score(y_test, y_pred)
    
    constraints_met = "YES" if recall_bankrupt >= 0.62 and specificity >= 0.70 else "NO"
    
    print(f"{name:<30} {acc:.4f}     {recall_bankrupt:.4f}               {specificity:.4f}               {constraints_met}")

Model                          Accuracy   Recall (Bankrupt)    Specificity (Healthy) Constraints Met
LDA                            0.9115     0.4516               0.9224               NO
Logistic Regression            0.9097     0.4839               0.9198               NO
KNN (k=5)                      0.8768     0.5000               0.8857               NO
KNN (k=7)                      0.8869     0.5161               0.8957               NO
KNN (k=9)                      0.8977     0.5484               0.9060               NO
Naïve Bayes                    0.8735     0.5323               0.8815               NO
SVM (RBF)                      0.9108     0.4839               0.9209               NO
MLP (64, 32)                   0.8511     0.4839               0.8598               NO


c:\Users\agenidounias\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:781: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (400) reached and the optimization hasn't converged yet.
  warnings.warn(


### Αξιολόγηση Αποτελεσμάτων

**Γενική Παρατήρηση:**
Παρατηρούμε ότι παρόλο που η **ακρίβεια (Accuracy)** είναι υψηλή σε όλα τα μοντέλα (85% - 91%), αυτό είναι παραπλανητικό λόγω της ανισορροπίας των κλάσεων στο Test Set (3:1). Τα μοντέλα τείνουν να προβλέπουν την πλειοψηφούσα κλάση (Healthy).

**Ανάλυση Περιορισμών:**
Κανένα από τα αρχικά μοντέλα δεν ικανοποίησε ταυτόχρονα τους περιορισμούς:
*   **Recall (Bankrupt) ≥ 62%**
*   **Specificity (Healthy) ≥ 70%**

Το καλύτερο Recall πέτυχε το **KNN (k=9)** με 54.84%, το οποίο όμως απέχει από το στόχο του 62%. Το Specificity είναι εξαιρετικά υψηλό σε όλα τα μοντέλα (>85%), που σημαίνει ότι τα μοντέλα είναι πολύ "συντηρητικά" στο να προβλέψουν χρεωκοπία.

**Επόμενο Βήμα:**
Έχουμε ήδη κάνει downsampling στο training set φέρνοντας την αναλογία στο **3:1** (Healthy:Bankrupt), θα δοκιμάσουμε να ενισχύσουμε περαιτέρω την ικανότητα εντοπισμού χρεωκοπιών χρησιμοποιώντας **βάρη**. Θα δώσουμε μεγαλύτερη "ποινή" στα λάθη που αφορούν τις χρεωκοπημένες επιχειρήσεις, ώστε να αναγκάσουμε τα μοντέλα να αυξήσουν το Recall τους, θυσιάζοντας πιθανώς λίγο από το πολύ υψηλό Specificity.

**Άσκηση 5 — Αξιολόγηση & Αρχείο Εξόδου**

Δημιουργήστε το αρχείο "classification_results.xlsx" στον φάκελο output. Για κάθε μοντέλο καταγράψτε τις εξής στήλες:
• Classifier Name
• Train/Test Set
• Samples Count
• TP (True Positives)
• TN
• FP
• FN
• Precision
• Recall
• F1 Score
• Accuracy

In [ ]:
# Define models with class weights
models_weighted = {
    "Logistic Regression (Balanced)": LogisticRegression(max_iter=1000, class_weight='balanced'),
    "SVM (Balanced)": SVC(kernel='rbf', class_weight='balanced'),
    "Logistic Regression (Custom 1:3)": LogisticRegression(max_iter=1000, class_weight={0: 1, 1: 3}),
    "SVM (Custom 1:3)": SVC(kernel='rbf', class_weight={0: 1, 1: 3}),
}

print("Training with Class Weights to meet constraints (Recall >= 62%, Specificity >= 70%)...\n")
print(f"{'Model':<35} {'Accuracy':<10} {'Recall (Bankrupt)':<20} {'Specificity (Healthy)':<20} {'Constraints Met'}")
print("="*105)

results = []

for name, model in models_weighted.items():
    model.fit(X_train, y_train)
    
    # Evaluate on both Train and Test sets
    for set_name, X_set, y_set in [("Train", X_train, y_train), ("Test", X_test, y_test)]:
        y_pred = model.predict(X_set)
        
        # Calculate metrics
        tn, fp, fn, tp = confusion_matrix(y_set, y_pred).ravel()
        precision = precision_score(y_set, y_pred)
        recall = recall_score(y_set, y_pred)
        f1 = f1_score(y_set, y_pred)
        acc = accuracy_score(y_set, y_pred)
        
        results.append({
            "Classifier Name": name,
            "Train/Test Set": set_name,
            "Samples Count": len(y_set),
            "TP (True Positives)": tp,
            "TN": tn,
            "FP": fp,
            "FN": fn,
            "Precision": precision,
            "Recall": recall,
            "F1 Score": f1,
            "Accuracy": acc
        })
        
        # Print only Test results for the console output requested previously
        if set_name == "Test":
            report = classification_report(y_set, y_pred, output_dict=True)
            recall_bankrupt = report['1']['recall']     
            specificity = report['0']['recall']         
            constraints_met = "YES" if recall_bankrupt >= 0.62 and specificity >= 0.70 else "NO"
            
            print(f"{name:<35} {acc:.4f}     {recall_bankrupt:.4f}               {specificity:.4f}               {constraints_met}")

# Create DataFrame and save to Excel
import os
if not os.path.exists('output'):
    os.makedirs('output')

results_df = pd.DataFrame(results)
results_df.to_excel('output/classification_results.xlsx', index=False)
print("\nResults saved to output/classification_results.xlsx")

Training with Class Weights to meet constraints (Recall >= 62%, Specificity >= 70%)...

Model                               Accuracy   Recall (Bankrupt)    Specificity (Healthy) Constraints Met
Logistic Regression (Balanced)      0.7704     0.7742               0.7703               YES
SVM (Balanced)                      0.7768     0.8065               0.7761               YES
Logistic Regression (Custom 1:3)    0.8645     0.5484               0.8720               NO
SVM (Custom 1:3)                    0.8343     0.6613               0.8384               YES
